File: backend/app/schemas/deployment.py
"""Pydantic schemas for deployment endpoints."""

from datetime import datetime
from typing import Optional
from uuid import UUID
from pydantic import BaseModel


class DeployRequest(BaseModel):
    """Request body for deploying an agent."""

    version_id: Optional[UUID] = None  # Uses current version if not specified


class DeploymentResponse(BaseModel):
    """Response for a single deployment."""

    id: UUID
    agent_id: UUID
    version_id: UUID
    status: str
    container_id: Optional[str] = None
    image_id: Optional[str] = None
    port: Optional[int] = None
    error_message: Optional[str] = None
    created_by: Optional[UUID] = None
    created_at: datetime
    started_at: Optional[datetime] = None
    stopped_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class DeploymentWithContainerResponse(DeploymentResponse):
    """Deployment response with container status info."""

    container: Optional[dict] = None


class DeploymentListResponse(BaseModel):
    """Response for listing deployments."""

    data: list[DeploymentResponse]
    total: int


class DeployResponse(BaseModel):
    """Response after initiating deployment."""

    deployment: DeploymentResponse
    message: str


class ChatRequest(BaseModel):
    """Request body for chat endpoint."""

    message: str
    conversation_id: Optional[str] = None


class ChatResponse(BaseModel):
    """Response from chat endpoint."""

    response: str
    conversation_id: str

File: frontend/src/types/deployment.ts
/**
 * Deployment-related types.
 */

/**
 * Deployment status values.
 */
export type DeploymentStatus =
  | 'pending'
  | 'building'
  | 'starting'
  | 'running'
  | 'stopping'
  | 'stopped'
  | 'failed';

/**
 * Container status info.
 */
export interface ContainerStatus {
  status: string;
  running: boolean;
  health: string;
  started_at?: string;
  finished_at?: string;
}

/**
 * Agent deployment record.
 */
export interface Deployment {
  id: string;
  agent_id: string;
  version_id: string;
  status: DeploymentStatus;
  container_id?: string;
  image_id?: string;
  port?: number;
  error_message?: string;
  created_by?: string;
  created_at: string;
  started_at?: string;
  stopped_at?: string;
  container?: ContainerStatus;
}

/**
 * Request body for deploying an agent.
 */
export interface DeployRequest {
  version_id?: string;
}

/**
 * Response from deploy endpoint.
 */
export interface DeployResponse {
  deployment: Deployment;
  message: string;
}

/**
 * Response from listing deployments.
 */
export interface DeploymentListResponse {
  data: Deployment[];
  total: number;
}

/**
 * Request body for chat endpoint.
 */
export interface ChatRequest {
  message: string;
  conversation_id?: string;
}

/**
 * Response from chat endpoint.
 */
export interface ChatResponse {
  response: string;
  conversation_id: string;
}

/**
 * Chat message in conversation.
 */
export interface ChatMessage {
  id: string;
  role: 'user' | 'assistant';
  content: string;
  timestamp: Date;
}

/**
 * WebSocket message types.
 */
export type WebSocketMessageType = 'message' | 'chunk' | 'done' | 'error';

/**
 * WebSocket message format.
 */
export interface WebSocketMessage {
  type: WebSocketMessageType;
  content?: string;
  conversation_id?: string;
  error?: string;
}
